core/shared/src/main/scala/zio/Scope.scala

```diff
@@ -209,12 +219,20 @@ object Scope {
    * Snoyman @snoyberg.
    * (https://github.com/snoyberg/conduit/blob/master/resourcet/Control/Monad/Trans/Resource/Internal.hs)
    */
-  private abstract class ReleaseMap extends Serializable {
+  private final class ReleaseMap extends Serializable {
 
-    /**
-     * An opaque identifier for a finalizer stored in the map.
-     */
-    type Key
+    // The sorting order of the LongMap uses bit ordering (000, 001, ... 111 but with 64 bits). This
+    // works out to be `0 ... Long.MaxValue, Long.MinValue, ... -1`. The order of the map is mainly
+    // important for the finalization, in which we want to walk it in reverse order. So we insert
+    // into the map using keys that will build it in reverse. That way, when we do the final iteration,
+    // the finalizers are already in correct order.
+    private[this] val ref =
+      Ref.unsafe.make[State](Running(-1L, LongMap.empty))(Unsafe)
```

Comment:
`ReleaseMap` could extend `AtomicRef` directly to avoid pointer chasing. Since it is private, that shouldn't be a problem.

Reply:
Nice, good point 👍